ConfigDispatcherFileWatcherTest always fails on Windows

These tests now always fail on Windows builds so it might be a real bug e.g. caused by #3004.

See this GHA build: https://github.com/wborn/openhab-core/actions/runs/4210674830/jobs/7308736469

### configurationFileCreated

```
[ERROR] Tests run: 6, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 1.562 s <<< FAILURE! - in org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest
[ERROR] org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileCreated  Time elapsed: 1.426 s  <<< FAILURE!
Wanted but not invoked:
configDispatcherMock.processConfigFile(
    myPath.cfg
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileCreated(ConfigDispatcherFileWatcherTest.java:50)

However, there was exactly 1 interaction with this mock:
configDispatcherMock.processConfigFile(
    conf\services
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcher.<init>(ConfigDispatcherFileWatcher.java:58)


	at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileCreated(ConfigDispatcherFileWatcherTest.java:50)
```

### configurationFileRemoved

```
[ERROR] org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileRemoved  Time elapsed: 0.004 s  <<< FAILURE!
Wanted but not invoked:
configDispatcherMock.fileRemoved(
    "D:\a\openhab-core\openhab-core\bundles\org.openhab.core.config.dispatch\myPath.cfg"
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileRemoved(ConfigDispatcherFileWatcherTest.java:82)

However, there was exactly 1 interaction with this mock:
configDispatcherMock.processConfigFile(
    conf\services
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcher.<init>(ConfigDispatcherFileWatcher.java:58)


	at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileRemoved(ConfigDispatcherFileWatcherTest.java:82)
```

### configurationFileModified

```
[ERROR] org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileModified  Time elapsed: 0.007 s  <<< FAILURE!
Wanted but not invoked:
configDispatcherMock.processConfigFile(
    myPath.cfg
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileModified(ConfigDispatcherFileWatcherTest.java:58)

However, there was exactly 1 interaction with this mock:
configDispatcherMock.processConfigFile(
    conf\services
);
-> at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcher.<init>(ConfigDispatcherFileWatcher.java:58)


	at org.openhab.core.config.dispatch.internal.ConfigDispatcherFileWatcherTest.configurationFileModified(ConfigDispatcherFileWatcherTest.java:58)
```

--- a/bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcher.java
+++ b/bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcher.java
@@ -79,7 +79,7 @@ public class ConfigDispatcherFileWatcher implements WatchService.WatchEventListe
                 configDispatcher.fileRemoved(fullPath.toString());
             }
         } catch (IOException e) {
-            logger.error("Failed to process watch event {} for {}", kind, path);
+            logger.error("Failed to process watch event {} for {}", kind, path, e);
         }
     }
 }

--- a/bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcher.java
+++ b/bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcher.java
@@ -12,7 +12,6 @@
  */
 package org.openhab.core.config.dispatch.internal;
 
-import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -65,19 +64,19 @@ public class ConfigDispatcherFileWatcher implements WatchService.WatchEventListe
 
     @Override
     public void processWatchEvent(WatchService.Kind kind, Path path) {
+        Path fullPath = watchService.getWatchPath().resolve(path);
         try {
             if (kind == WatchService.Kind.CREATE || kind == WatchService.Kind.MODIFY) {
-                if (!Files.isHidden(path) && path.toString().endsWith(".cfg")) {
-                    configDispatcher.processConfigFile(path.toFile());
+                if (!Files.isHidden(fullPath) && fullPath.toString().endsWith(".cfg")) {
+                    configDispatcher.processConfigFile(fullPath.toFile());
                 }
             } else if (kind == WatchService.Kind.DELETE) {
                 // Detect if a service specific configuration file was removed. We want to
                 // notify the service in this case with an updated empty configuration.
-                File configFile = path.toFile();
-                if (Files.isHidden(path) || Files.isDirectory(path) || !path.toString().endsWith(".cfg")) {
+                if (Files.isHidden(fullPath) || Files.isDirectory(fullPath) || !fullPath.toString().endsWith(".cfg")) {
                     return;
                 }
-                configDispatcher.fileRemoved(configFile.getAbsolutePath());
+                configDispatcher.fileRemoved(fullPath.toString());
             }
         } catch (IOException e) {
             logger.error("Failed to process watch event {} for {}", kind, path);

--- a/bundles/org.openhab.core.config.dispatch/src/test/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcherTest.java
+++ b/bundles/org.openhab.core.config.dispatch/src/test/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcherTest.java
@@ -14,12 +14,15 @@ package org.openhab.core.config.dispatch.internal;
 
 import static org.mockito.Mockito.*;
 
+import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 import org.openhab.core.service.WatchService;
@@ -36,59 +39,58 @@ public class ConfigDispatcherFileWatcherTest {
     private @Mock @NonNullByDefault({}) ConfigDispatcher configDispatcherMock;
     private @Mock @NonNullByDefault({}) WatchService watchService;
 
+    private @TempDir @NonNullByDefault({}) Path tempDir;
+
+    private @NonNullByDefault({}) Path cfgPath;
+    private @NonNullByDefault({}) Path nonCfgPath;
+
     @BeforeEach
-    public void setUp() {
+    public void setUp() throws IOException {
         configDispatcherFileWatcher = new ConfigDispatcherFileWatcher(configDispatcherMock, watchService);
         verify(configDispatcherMock).processConfigFile(any());
 
-        when(watchService.getWatchPath()).thenReturn(Path.of("").toAbsolutePath());
+        when(watchService.getWatchPath()).thenReturn(tempDir.toAbsolutePath());
+
+        cfgPath = tempDir.resolve("myPath.cfg");
+        nonCfgPath = tempDir.resolve("myPath");
+
+        Files.createFile(cfgPath);
+        Files.createFile(nonCfgPath);
     }
 
     @Test
-    public void configurationFileCreated() {
-        Path path = Path.of("myPath.cfg");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.CREATE, path);
-
-        verify(configDispatcherMock).processConfigFile(path.toAbsolutePath().toFile());
+    public void configurationFileCreated() throws IOException {
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.CREATE, cfgPath);
+        verify(configDispatcherMock).processConfigFile(cfgPath.toAbsolutePath().toFile());
     }
 
     @Test
-    public void configurationFileModified() {
-        Path path = Path.of("myPath.cfg");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.MODIFY, path);
-
-        verify(configDispatcherMock).processConfigFile(path.toAbsolutePath().toFile());
+    public void configurationFileModified() throws IOException {
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.MODIFY, cfgPath);
+        verify(configDispatcherMock).processConfigFile(cfgPath.toAbsolutePath().toFile());
     }
 
     @Test
     public void nonConfigurationFileCreated() {
-        Path path = Path.of("myPath");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.CREATE, path);
-
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.CREATE, nonCfgPath);
         verifyNoMoreInteractions(configDispatcherMock);
     }
 
     @Test
     public void nonConfigurationFileModified() {
-        Path path = Path.of("myPath");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.MODIFY, path);
-
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.MODIFY, nonCfgPath);
         verifyNoMoreInteractions(configDispatcherMock);
     }
 
     @Test
     public void configurationFileRemoved() {
-        Path path = Path.of("myPath.cfg");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.DELETE, path);
-
-        verify(configDispatcherMock).fileRemoved(path.toAbsolutePath().toString());
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.DELETE, cfgPath);
+        verify(configDispatcherMock).fileRemoved(cfgPath.toAbsolutePath().toString());
     }
 
     @Test
     public void nonConfigurationFileRemoved() {
-        Path path = Path.of("myPath");
-        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.DELETE, path);
-
+        configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.DELETE, nonCfgPath);
         verifyNoMoreInteractions(configDispatcherMock);
     }
 }

--- a/bundles/org.openhab.core.config.dispatch/src/test/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcherTest.java
+++ b/bundles/org.openhab.core.config.dispatch/src/test/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherFileWatcherTest.java
@@ -40,6 +40,8 @@ public class ConfigDispatcherFileWatcherTest {
     public void setUp() {
         configDispatcherFileWatcher = new ConfigDispatcherFileWatcher(configDispatcherMock, watchService);
         verify(configDispatcherMock).processConfigFile(any());
+
+        when(watchService.getWatchPath()).thenReturn(Path.of("").toAbsolutePath());
     }
 
     @Test
@@ -47,7 +49,7 @@ public class ConfigDispatcherFileWatcherTest {
         Path path = Path.of("myPath.cfg");
         configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.CREATE, path);
 
-        verify(configDispatcherMock).processConfigFile(path.toFile());
+        verify(configDispatcherMock).processConfigFile(path.toAbsolutePath().toFile());
     }
 
     @Test
@@ -55,7 +57,7 @@ public class ConfigDispatcherFileWatcherTest {
         Path path = Path.of("myPath.cfg");
         configDispatcherFileWatcher.processWatchEvent(WatchService.Kind.MODIFY, path);
 
-        verify(configDispatcherMock).processConfigFile(path.toFile());
+        verify(configDispatcherMock).processConfigFile(path.toAbsolutePath().toFile());
     }
 
     @Test

